lib/parser/canonicalize-omp.cc

```diff
@@ -0,0 +1,129 @@
+          // OpenMPLoopConstruct
+          if (auto *ompLoop{
+                  std::get_if<OpenMPLoopConstruct>(&ompCons->value().u)}) {
+            RewriteOpenMPLoopConstruct(ompLoop, block, it);
+          }
+        } else if (auto *endDir{
+                       std::get_if<common::Indirection<OpenMPEndLoopDirective>>(
+                           &exec->u)}) {
+          // Unmatched OpenMPEndloopdirective
+          messages_.Say(endDir->value().source,
+              "The %s directive must follow the DO loop associated with the "
+              "loop construct"_err_en_US,
+              parser::ToUpperCaseLetters(endDir->value().source.ToString()));
+        }
+      }
+    }  // Block list
+  }
+
+private:
+  void RewriteOpenMPLoopConstruct(
+      OpenMPLoopConstruct *x, Block &block, Block::iterator it) {
+    // Check the sequence of DoConstruct and OpenMPEndLoopDirective
+    // in the same iteration
+    //
+    // Original:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //   ExecutableConstruct -> DoConstruct
+    //   ExecutableConstruct -> OpenMPEndLoopDirective (if available)
+    //
+    // After rewriting:
+    //   ExecutableConstruct -> OpenMPConstruct -> OpenMPLoopConstruct
+    //     OmpLoopDirective
+    //     OmpClauseList
+    //     DoConstruct
+    //     OpenMPEndLoopDirective (if available)
+    Block::iterator nextIt;
+    OpenMPLoopConstruct *matchedLoopConstruct{nullptr};
+
+    auto &dir{std::get<OmpLoopDirective>(x->t)};
+    nextIt = it;
+    if (++nextIt != block.end()) {
+      if (auto *execNext{std::get_if<ExecutableConstruct>(&nextIt->u)}) {
+        if (auto *doCons{
+                std::get_if<common::Indirection<DoConstruct>>(&execNext->u)}) {
+          const auto &loopControl{doCons->value().GetLoopControl()};
+          if (loopControl.has_value()) {
+            matchedLoopConstruct = x;
+            // move DoConstruct
+            std::get<std::optional<DoConstruct>>(matchedLoopConstruct->t) =
+                std::move(doCons->value());
+            nextIt = block.erase(nextIt);
+
+            // try to match OpenMPEndLoopDirective
+            if (auto *execEnd{std::get_if<ExecutableConstruct>(&nextIt->u)}) {
+              if (auto *endDir{
+                      std::get_if<common::Indirection<OpenMPEndLoopDirective>>(
+                          &execEnd->u)}) {
+                std::get<std::optional<OpenMPEndLoopDirective>>(
+                    matchedLoopConstruct->t) = std::move(endDir->value());
+                nextIt = block.erase(nextIt);
+              }
+            }
+          } else {
+            messages_.Say(dir.source,
+                "DO loop after the %s directive "
+                "must have loop control"_err_en_US,
+                parser::ToUpperCaseLetters(dir.source.ToString()));
+            return;
+          }
+        }
+      }
+    }
+    if (!matchedLoopConstruct) {
```

Comment:
I don't think you need `matchedLoopConstruct`. Where you use it up above you can use `x` instead. And if you return after finding the `DoConstruct` you can unconditionally report the error here.

Reply:
Thanks, I've got rid of it.